src/transformers/models/chameleon/modeling_chameleon.py

```diff
@@ -0,0 +1,2061 @@
+    """
+    Chameleon flash attention module. This module inherits from `ChameleonAttention` as the weights of the module stays
+    untouched. The only required change would be on the forward pass where it needs to correctly call the public API of
+    flash attention and deal with padding tokens in case the input contains any of them.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # TODO: Should be removed once Flash Attention for RoCm is bumped to 2.1.
+        # flash_attn<2.1 generates top-left aligned causal mask, while what is needed here is bottom-right alignement, that was made default for flash_attn>=2.1. This attribute is used to handle this difference. Reference: https://github.com/Dao-AILab/flash-attention/releases/tag/v2.1.0.
+        # Beware that with flash_attn<2.1, using q_seqlen != k_seqlen (except for the case q_seqlen == 1) produces a wrong mask (top-left).
+        self._flash_attn_uses_top_left_mask = not is_flash_attn_greater_or_equal_2_10()
+
+    # Ignore copy
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.LongTensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Cache] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        **kwargs,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        output_attentions = False
+
+        bsz, q_len, _ = hidden_states.size()
+
+        query_states = self.q_proj(hidden_states)
+        key_states = self.k_proj(hidden_states)
+        value_states = self.v_proj(hidden_states)
+
+        if self.model_parallel_size == 1:
+            query_states = query_states.reshape(-1, 1, self.head_dim)
+            query_states = query_states.tile((1, self.num_heads, 1))
+            query_states = self.q_norm(query_states)
+            query_states = query_states[:, 0, :]
+
+            key_states = key_states.reshape(-1, 1, self.head_dim)
+            key_states = key_states.tile((1, self.num_key_value_heads, 1))
+            key_states = self.k_norm(key_states)
+            key_states = key_states[:, 0, :]
+
+        elif self.model_parallel_size > 1:
+            query_states = query_states.reshape(-1, self.num_heads, self.head_dim)
+            query_states = query_states.tile((1, self.model_parallel_size, 1))
+            query_states = query_states.reshape(-1, self.model_parallel_size, self.num_heads, self.head_dim).transpose(
+                1, 2
+            )
+            query_states = query_states.reshape(-1, self.model_parallel_size * self.num_heads, self.head_dim)
+            query_states = self.q_norm(query_states)
+            query_states = query_states.reshape(-1, self.model_parallel_size, self.head_dim)
+            query_states = query_states[:, 0, :]
```

Comment:
I got to decrease the norm matrix by 4 on the first dim, and now we don't have to tile before forwarding. But we still have the tile and expansion on 7B model because that one was not sharded and applying the same calculations by just repeating weight/bias `num_heads` times does not work

Regarding the ROPE, for me it seems like a hard task to shift norm and QKV weights to match the distribution. I did some experiments, but since the QKV proj is a matmul, simply doing the same transformation of norm alphas will not work

Reply:
What do you think about this?

```python
if self.model_parallel_size == 1:
            query_states = query_states.reshape(-1, 1, self.head_dim)
            query_states = query_states.tile((1, self.num_heads, 1))
            query_states = self.q_norm(query_states)
            query_states = query_states[:, 0, :]

            key_states = key_states.reshape(-1, 1, self.head_dim)
            key_states = key_states.tile((1, self.num_key_value_heads, 1))
            key_states = self.k_norm(key_states)
            key_states = key_states[:, 0, :]
        else:
            query_states = query_states.reshape(-1, self.num_heads, self.head_dim)
            query_states = self.q_norm(query_states)

            key_states = key_states.reshape(-1, self.num_key_value_heads, self.head_dim)
            key_states = self.k_norm(key_states)
```

Reply:
a LOT better, don't know why you hacve to reshape -1, but this is already really good for model paralle_size != 1 buty there is still the rope permutation that you are doing afterward

Reply:
I just needed to match the last dims to shape of multiplications, because we expand norm shape to `(head, head_dim)` for sharded model's peculiarity of applying different alphas to different heads

Reply:
using F.layer_norm with dim could also make this simpler with a custom ChameleonLayerNorm (the same as https://github.com/huggingface/transformers/blob/3d6e81be3bf09a4d7d7886dffe557f734ac416d9/src/transformers/models/olmo/modeling_olmo.py#L82 )